orc/src/test/java/org/apache/iceberg/orc/TestOrcMetrics.java

```diff
@@ -0,0 +1,138 @@
+ */
+public class TestOrcMetrics extends TestMetrics {
+
+  static final ImmutableSet<Object> BINARY_TYPES = ImmutableSet.of(Type.TypeID.BINARY,
+      Type.TypeID.FIXED, Type.TypeID.UUID);
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Override
+  public FileFormat fileFormat() {
+    return FileFormat.ORC;
+  }
+
+  @Override
+  public Metrics getMetrics(InputFile file) {
+    return OrcMetrics.fromInputFile(file);
+  }
+
+  @Override
+  public File writeRecords(Schema schema, GenericData.Record... records) throws IOException {
+    return OrcWritingTestUtils.writeRecords(temp, schema, records);
+  }
+
+  @Override
+  public File writeRecords(Schema schema, Map<String, String> properties,
+                           GenericData.Record... records) throws IOException {
+    return OrcWritingTestUtils.writeRecords(temp, schema, properties, records);
+  }
+
+  private boolean isBinaryType(Type type) {
+    return BINARY_TYPES.contains(type.typeId());
+  }
+
+  @Override
+  protected <T> void assertBounds(int fieldId, Type type, T lowerBound, T upperBound, Metrics metrics) {
+    if (isBinaryType(type)) {
+      return;
+    }
+
+    Map<Integer, ByteBuffer> lowerBounds = metrics.lowerBounds();
+    Map<Integer, ByteBuffer> upperBounds = metrics.upperBounds();
+
+    Assert.assertEquals(
+        lowerBound,
+        lowerBounds.containsKey(fieldId) ? fromByteBuffer(type, lowerBounds.get(fieldId)) : null);
+    Assert.assertEquals(
+        upperBound,
+        upperBounds.containsKey(fieldId) ? fromByteBuffer(type, upperBounds.get(fieldId)) : null);
+  }
+
+  @Test
+  public void testMetricsForListAndMapElements() throws IOException {
+    Types.StructType structType = Types.StructType.of(
+        required(1, "leafIntCol", Types.IntegerType.get()),
+        optional(2, "leafStringCol", Types.StringType.get())
+    );
+    Schema schema = new Schema(
+        optional(3, "intListCol", Types.ListType.ofRequired(4, Types.IntegerType.get())),
+        optional(5, "mapCol", Types.MapType.ofRequired(6, 7, Types.StringType.get(), structType))
+    );
+
+    GenericData.Record record = new GenericData.Record(AvroSchemaUtil.convert(schema.asStruct()));
+    record.put("intListCol", Lists.newArrayList(10, 11, 12));
+    GenericData.Record struct = new GenericData.Record(AvroSchemaUtil.convert(structType));
+    struct.put("leafIntCol", 1);
+    struct.put("leafStringCol", "BBB");
+    Map<String, GenericData.Record> map = Maps.newHashMap();
+    map.put("4", struct);
+    record.put(1, map);
+
+    File recordsFile = writeRecords(schema, record);
+
+    Metrics metrics = getMetrics(Files.localInput(recordsFile));
+    Assert.assertEquals(1L, (long) metrics.recordCount());
+    assertCounts(1, 1, 0, metrics);
+    assertBounds(1, Types.IntegerType.get(), 1, 1, metrics);
+    assertCounts(2, 1, 0, metrics);
+    assertBounds(2, Types.StringType.get(),
+        CharBuffer.wrap("BBB"), CharBuffer.wrap("BBB"), metrics);
```

Comment:
If this excludes bounds for repeated columns, then the assertions here will match the ones for Parquet, so both could be moved back into the base test.